cla_backend/libs/eligibility_calculator/calculator.py

```diff
@@ -390,24 +399,64 @@ def _translate_case(self):
-        if hasattr(self.case_data, "property_data"):
-            request_data.update(translate_property(self.case_data.property_data))
-        if hasattr(self.case_data.you, "income") and hasattr(self.case_data.you, "deductions"):
-            request_data.update(translate_employment(self.case_data.you.income, self.case_data.you.deductions))
-        if hasattr(self.case_data.you, "income"):
-            request_data.update(translate_income(self.case_data.you.income))
-        if hasattr(self.case_data, "facts"):
-            request_data['applicant'].update(translate_age(submission_date, self.case_data.facts))
+        if hasattr(case_data, "facts"):
+            request_data['applicant'].update(EligibilityChecker._applicant_data(submission_date, case_data.facts))
+        if hasattr(case_data, "facts"):
+            request_data.update(translate_dependants(submission_date, case_data.facts))
+
+        request_data.update(EligibilityChecker._capital_data(case_data))
+        if not EligibilityChecker._complete_cfe_capital_data(request_data):
+            request_data['assessment']['section_capital'] = 'incomplete'
+
+        request_data.update(EligibilityChecker._income_data(case_data))
+        if not EligibilityChecker._complete_cfe_income_data(request_data):
+            request_data['assessment']['section_gross_income'] = 'incomplete'
+
+        return request_data
+
+    @staticmethod
+    def _income_data(case_data):
+        request_data = {}
+        if hasattr(case_data.you, "income") and hasattr(case_data.you, "deductions"):
+            request_data.update(translate_employment(case_data.you.income, case_data.you.deductions))
+        if hasattr(case_data.you, "income"):
+            request_data.update(translate_income(case_data.you.income))
         return request_data
 
+    @staticmethod
+    def _applicant_data(submission_date, facts):
+        request_data = {}
+        request_data.update(translate_age(submission_date, facts))
+        request_data.update(translate_applicant(facts))
+
+        return request_data
+
+    @staticmethod
+    def _capital_data(case_data):
+        request_data = {}
+        if hasattr(case_data.you, "savings"):
+            request_data.update(translate_savings(case_data.you.savings))
+
+        if hasattr(case_data, "property_data"):
+            request_data.update(translate_property(case_data.property_data))
+        return request_data
+
+    @staticmethod
+    def _complete_applicant_data(request_data, facts):
+        return "dependants" in request_data and hasattr(facts, "has_partner")
+
+    @staticmethod
+    def _complete_cfe_income_data(request_data):
+        employment_data = "employment_details" in request_data or "self_employment_details" in request_data
+        income_data = "regular_transactions" in request_data
+        return employment_data and income_data
+
+    @staticmethod
+    def _complete_cfe_capital_data(request_data):
```

Comment:
I don't think this should look at the CFE request, where the schema is clear that these fields are optional. The way we've written the translators happens to put in a blank "properties" section, and others, but that doesn't seem something we should rely on here.

Instead the CaseData object absolutely does define the non-presence of property answers to mean that that the question has not been asked, so I think we should use that.

Reply:
My intention is/was that this decision should be 'inside' the translators, so that they produce no output when invalid data is given. Agree the intent could be communicated clearer - I wasn't sure how to indicate that e.g. the 'regular_transactions' key comes from the 'income' translator (and so hence is gross income based) and the 'capitals' key comes from the 'savings' translator and hence is from the 'capital' section

Reply:
> e.g. the 'regular_transactions' key comes from the 'income' translator (and so hence is gross income based)

In CFE's schema, `regular_transactions` is for both incomes, like benefits, but it's also for outgoings of types: child_care, rent_or_mortgage, maintenance_out, legal_aid. Whilst CLA doesn't happen to capture those outgoings at the current time, it just makes me confused to see the presence of 'regular_transactions' being used to mean "income is completely filled in on the form". It's flakey. Yes the intent is clearer with CFE_INCOME_KEY. But CLA's native way to tell us if "income is completely filled in in the form" is to check the case_data object for presence of keys. That seems much more straightforward than depending on the output of potentially intertwined translators. But perhaps I'm missing something here.

Reply:
I wanted the translators to have this knowledge (which they have to have, as they need to check for existence of properties in general) rather than put it somewhere else. It seemed logical for the translator to not create the key when the fields are not fully populated, and then check for presence afterwards. I've added a tighter abstraction in the last commit to make it clearer - maybe it needs more comments explaining the idea?

Reply:
yes this really helps. let's go with this approach 👍